service/matching/physical_task_queue_manager.go

```diff
@@ -128,12 +134,108 @@ type (
+}
+
+// returns the total number of tasks in the buffer
+func (cb *circularTaskBuffer) totalTasks() int {
+	totalTasks := 0
+	for _, count := range cb.buffer {
+		totalTasks += count
+	}
+	return totalTasks
+}
+
+type taskTracker struct {
+	mutex             sync.RWMutex
+	clock             clock.TimeSource
+	startTime         time.Time     // time when taskTracker was initialized
+	startIntervalTime time.Time     // the starting time of a window in the buffer
+	interval          time.Duration // the duration of each mini- in the buffer
+	totalIntervalSize int           // the number over which rate of tasks added/dispatched would be determined
+	tasksInInterval   *circularTaskBuffer
+}
+
+func newTaskTracker(timeSource clock.TimeSource, intervalSize int, totalIntervalSize int) *taskTracker {
+	return &taskTracker{
+		clock:             timeSource,
+		startTime:         timeSource.Now(),
+		startIntervalTime: timeSource.Now(),
+		interval:          time.Duration(intervalSize) * time.Second, // Todo: Shivam - replace with config value
+		totalIntervalSize: totalIntervalSize,
+		tasksInInterval:   newCircularTaskBuffer((totalIntervalSize / intervalSize) + 1), // Todo: Shivam - replace hardcoded value with number of buckets
+	}
+}
+
+// advanceAndResetTracker is a helper to advance the trackers position and clear out any expired intervals
+func (s *taskTracker) advanceAndResetTracker(elapsed time.Duration) {
+	// Calculate the number of intervals elapsed since the start interval time
+	intervalsElapsed := int(elapsed / s.interval)
+
+	if intervalsElapsed != 0 {
+		for i := 0; i < intervalsElapsed; i++ {
+			s.tasksInInterval.advance() // advancing our circular buffer's position until we land on the right interval
+		}
+		s.startIntervalTime = s.startIntervalTime.Add(time.Duration(intervalsElapsed) * s.interval)
+	}
+}
+
+// trackTasks is responsible for adding/removing tasks from the current time that falls in the appropriate interval
+func (s *taskTracker) trackTasks() {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	currentTime := s.clock.Now()
+
+	// Calculate elapsed time from the latest start interval time
+	elapsed := currentTime.Sub(s.startIntervalTime)
+	s.advanceAndResetTracker(elapsed)
+	s.tasksInInterval.incrementTaskCount()
+}
+
+// rate is responsible for returning the rate of tasks added/dispatched in a given interval
+func (s *taskTracker) rate() float32 {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	currentTime := s.clock.Now()
+
+	// Calculate elapsed time from the latest start interval time
+	elapsed := currentTime.Sub(s.startIntervalTime)
+	s.advanceAndResetTracker(elapsed)
+	totalTasks := s.tasksInInterval.totalTasks()
+
+	// time passed since start of the current window + totalIntervalSize in milliseconds
+	elapsedTime := currentTime.Sub(s.startIntervalTime).Milliseconds() + int64(s.totalIntervalSize*1000)
+	// rate per second
+	return (float32(totalTasks) / float32(elapsedTime)) * 1000
```

Comment:
I was thinking of having `s.totalIntervalSize` as a duration/time object to remove the conversion here but decided against it:

- I need the int value of `s.totalIntervalSize` since we calculate `tasksInInterval` based on that and creating another new variable for the structure, just for this, did not make sense.

Let me know if you think otherwise.

Reply:
I have changed `intervalSize` to `bucketSize`, and `startIntervalTime` to `bucketStartTime` for better readability.